tempesta_fw/http_parser.c

```diff
@@ -2602,168 +2723,252 @@ __resp_parse_expires(TfwHttpResp *resp, unsigned char *data, size_t len)
 		if (__fsm_n == CSTR_POSTPONE)
 			tfw_http_msg_hdr_chunk_fixup(msg, p, __fsm_sz);
 		if (__fsm_n < 0)
 			return __fsm_n;
-		resp->expires += parser->_acc * 60;
+		parser->_date += parser->_acc * 60;
 		parser->_acc = 0;
-		__FSM_I_MOVE_n(Resp_I_ExpSecCln, __fsm_n);
+		__FSM_I_MOVE_n(Resp_I_DateSecCln, __fsm_n);
 	}
 
-	__FSM_STATE(Resp_I_ExpSecCln) {
+	__FSM_STATE(Resp_I_DateSecCln) {
 		if (likely(c == ':'))
-			__FSM_I_MOVE(Resp_I_ExpSec);
+			__FSM_I_MOVE(Resp_I_DateSec);
 		return CSTR_NEQ;
 	}
 
-	__FSM_STATE(Resp_I_ExpSec) {
+	__FSM_STATE(Resp_I_DateSec) {
 		__fsm_sz = __data_remain(p);
 		__fsm_n = parse_int_ws(p, __fsm_sz, &parser->_acc);
 		if (__fsm_n == CSTR_POSTPONE)
 			tfw_http_msg_hdr_chunk_fixup(msg, p, __fsm_sz);
 		if (__fsm_n < 0)
 			return __fsm_n;
-		resp->expires += parser->_acc;
+		parser->_date += parser->_acc;
 		parser->_acc = 0;
-		__FSM_I_MOVE_n(Resp_I_EoL, __fsm_n);
+		__FSM_I_MOVE_n(Resp_I_DateSecSP, __fsm_n);
+	}
+
+	__FSM_STATE(Resp_I_DateSecSP) {
+		if (likely(isspace(c)))
+			__FSM_I_MOVE(Resp_I_DateZone);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Resp_I_DateZone) {
+		TRY_STR("GMT", Resp_I_EoL);
+		TRY_STR_INIT();
+		return CSTR_NEQ;
 	}
 
 	__FSM_STATE(Resp_I_EoL) {
-		/* Skip rest of line: ' GMT'. */
+		/* Skip the rest of the line. */
 		__fsm_sz = __data_remain(p);
 		__fsm_ch = memchreol(p, __fsm_sz);
-		if (__fsm_ch)
-			return __data_offset(__fsm_ch);
-		__FSM_I_MOVE_n(Resp_I_EoL, __fsm_sz);
+		if (!__fsm_ch)
+			__FSM_I_MOVE_n(Resp_I_EoL, __fsm_sz);
+		switch (parser->state) {
+		case Resp_HdrExpiresV:
+			resp->cache_ctl.expires = parser->_date;
+			resp->cache_ctl.flags |= TFW_HTTP_CC_HDR_EXPIRES;
+			break;
+		case Resp_HdrDateV:
+			resp->date = parser->_date;
+			break;
+		default:
+			TFW_DBG2("%s: Unknown caller's FSM state: [%d]\n",
+				 __func__, parser->state);
+			BUG();
+			return CSTR_NEQ;
+		}
+		return __data_offset(__fsm_ch);
+	}
+
+	} /* FSM END */
+done:
+	return r;
+}
+
+/*
+ * The value of "Expires:" header field is a date in HTTP-Date format.
+ * However, if the format of a date is invalid, that is interpreted
+ * as representing a time in the past (i.e., "already expired").
+ * See RFC 7234 5.3.
+ */
+static int
+__resp_parse_expires(TfwHttpResp *resp, unsigned char *data, size_t len)
+{
+	int ret = __resp_parse_http_date(resp, data, len);
+	if (ret < CSTR_POSTPONE) {  /* (ret < 0) && (ret != POSTPONE) */
+		BUG_ON(resp->parser.state != Resp_HdrExpiresV);
+		/*
+		 * On error just swallow the rest of the line.
+		 * @resp->expires stays zero - already expired.
+		 */
+		resp->parser._date = 0;
+		resp->parser._i_st = Resp_I_EoL;
+		ret = __resp_parse_http_date(resp, data, len);
+	}
+	return ret;
+}
+
+static int
+__resp_parse_age(TfwHttpResp *resp, unsigned char *data, size_t len)
+{
+	int r = CSTR_NEQ;
+	__FSM_DECLARE_VARS(resp);
+
+	__FSM_START(parser->_i_st) {
+
+	__FSM_STATE(Resp_I_Age) {
+		/* Eat OWS before the node ID. */
+		if (unlikely(IS_WS(c)))
```

Comment:
No need to check for `IS_WS(c)` here as `RGen_LWS` eats all the spaces before the token.